File: src/mgen/domain/sediment_ingest.py
"""Anti-corruption layer for the Aquatic Monitoring Database Sediment sheet.

Reads the "Sediment" sheet and normalises quirky header names, derives
the output Period and Season columns from source values, and strips
whitespace from site codes. Shared by both sediment and sediment_size
domain modules.

Source layout quirks:
  - First column header is a literal space character (' ')
  - Site column has a trailing space ('Site ')
  - Season values include "Additional - *" prefixes for incident monitoring
"""

from __future__ import annotations

import logging
from pathlib import Path

import pandas as pd

from mgen.shared.errors import DomainResult, ValidationError

__all__ = [
    "GRAIN_SIZE_SOURCE_COLUMNS",
    "SOURCE_SAM1_COL",
    "SOURCE_SAM3_COL",
    "IngestError",
    "make_error_result",
    "read_sediment_sheet",
]

logger = logging.getLogger(__name__)

# Source column names (exact strings from the workbook).
_SOURCE_PERIOD_COL = " "
_SOURCE_SITE_COL = "Site "
_SOURCE_SEASON_COL = "Season"
_SOURCE_DATE_COL = "Date"
SOURCE_SAM1_COL = "SAM1 %fine cover"
SOURCE_SAM3_COL = "SAM3 (%Fine Cover)"

# Grain-size source columns in output order.
GRAIN_SIZE_SOURCE_COLUMNS = [
    "Clay/silt (<0.06 mm)",
    "Sand (>0.06-2 mm)",
    "Small gravel (>2-8 mm)",
    "Small-med gravel (>8-16 mm)",
    "Med-large gravel (>16-32 mm)",
    "Large gravel (>32-64 mm)",
    "Small cobble (>64-128 mm)",
    "Large cobble (>128-256 mm)",
    "Boulders (>256 mm)",
    "Bedrock",
]

# Minimum required source columns for reading the sheet at all.
_REQUIRED_COLUMNS = frozenset(
    {_SOURCE_PERIOD_COL, _SOURCE_SITE_COL, _SOURCE_SEASON_COL, _SOURCE_DATE_COL}
)


class IngestError(Exception):
    """Raised when the Sediment sheet cannot be parsed."""


def read_sediment_sheet(path: Path) -> pd.DataFrame:
    """Read and normalise the Sediment sheet from the Aquatic Monitoring Database.

    Returns a DataFrame with normalised columns:
      Site, Date, Period, Season + SAM columns + grain-size columns.

    Raises:
        IngestError: If the file or sheet cannot be read or is missing
            required columns.
    """
    try:
        df = pd.read_excel(path, sheet_name="Sediment")
    except FileNotFoundError as exc:
        msg = f"File not found: {path}"
        raise IngestError(msg) from exc
    except ValueError as exc:
        msg = f"Sheet 'Sediment' not found in {path}"
        raise IngestError(msg) from exc
    except Exception as exc:
        msg = f"Failed to read {path}: {exc}"
        raise IngestError(msg) from exc

    # Validate required columns exist.
    missing = _REQUIRED_COLUMNS - set(df.columns)
    if missing:
        msg = f"Missing required columns: {sorted(missing)}"
        raise IngestError(msg)

    # Derive output Period and Season (vectorized).
    period_raw = df[_SOURCE_PERIOD_COL].astype(str).str.strip()
    season_raw = df[_SOURCE_SEASON_COL].astype(str).str.strip()

    df["Period"] = "Routine Construction"
    df.loc[period_raw == "Baseline", "Period"] = "Baseline"
    df.loc[season_raw.str.startswith("Additional"), "Period"] = "Incident"

    df["Season"] = "Spring"
    df.loc[season_raw.str.contains("Summer", na=False), "Season"] = "Summer"

    # Normalise Site (strip trailing whitespace).
    df["Site"] = df[_SOURCE_SITE_COL].astype(str).str.strip()

    # Ensure Date is datetime64[ns].
    df["Date"] = pd.to_datetime(df[_SOURCE_DATE_COL]).astype("datetime64[ns]")

    return df


def make_error_result(domain: str, path: Path, message: str) -> DomainResult:
    """Create a failed DomainResult with a single validation error."""
    return DomainResult(
        data=None,
        errors=[
            ValidationError(
                domain=domain,
                severity="error",
                file=str(path),
                sheet="Sediment",
                location="sheet",
                message=message,
            )
        ],
    )

File: src/r/data/domains/sediment_ingest.R
# sediment_ingest.R — ACL for the Aquatic Monitoring Database "Sediment" sheet.

SOURCE_SAM1_COL <- "SAM1 %fine cover"
SOURCE_SAM3_COL <- "SAM3 (%Fine Cover)"
GRAIN_SIZE_SOURCE_COLUMNS <- c(
  "Clay/silt (<0.06 mm)", "Sand (>0.06-2 mm)", "Small gravel (>2-8 mm)",
  "Small-med gravel (>8-16 mm)", "Med-large gravel (>16-32 mm)",
  "Large gravel (>32-64 mm)", "Small cobble (>64-128 mm)",
  "Large cobble (>128-256 mm)", "Boulders (>256 mm)", "Bedrock"
)

# Column identifiers as returned by readxl::read_excel(..., .name_repair="minimal").
# The source workbook's period header is a single space; readxl strips it to "".
# The source workbook's site header is "Site " (trailing space); readxl strips to "Site".
# The "" column cannot be accessed via df[[""]] (R returns NULL); use .sed_get_period().
.sed_period_col <- ""
.sed_site_col <- "Site"
.sed_season_col <- "Season"
.sed_date_col <- "Date"
# Required cols for validation — "" is checked via %in% names(df).
.sed_required <- c(.sed_period_col, .sed_site_col, .sed_season_col, .sed_date_col)

# Access the empty-named period column by position (df[[""]] returns NULL in R).
.sed_get_period <- function(df) df[[match("", names(df))]]

sediment_make_error <- function(domain, path, message) {
  DomainResult$new(data = NULL, errors = list(ValidationError(
    domain = domain, severity = "error", file = as.character(path),
    sheet = "Sediment", location = "sheet", message = message)))
}

read_sediment_sheet <- function(path) {
  df <- tryCatch(readxl::read_excel(path, sheet = "Sediment", .name_repair = "minimal"),
                 error = function(e) stop(sprintf(
                   "Sheet 'Sediment' not found or unreadable in %s", path)))
  missing <- setdiff(.sed_required, names(df))
  if (length(missing) > 0) {
    stop(sprintf("Missing required columns: %s",
      paste0("[", paste(sprintf("'%s'", sort(missing)), collapse = ", "), "]")))
  }
  period_raw <- trimws(as.character(.sed_get_period(df)))
  season_raw <- trimws(as.character(df[[.sed_season_col]]))

  Period <- rep("Routine Construction", nrow(df))
  Period[period_raw == "Baseline"] <- "Baseline"
  Period[startsWith(season_raw, "Additional")] <- "Incident"

  Season <- rep("Spring", nrow(df))
  Season[grepl("Summer", season_raw, fixed = TRUE)] <- "Summer"

  # Match pandas df[col].astype(str).str.strip(): a missing Site cell becomes
  # the literal string "nan" (these rows are not dropped downstream), so map
  # NA -> "nan" before trimming rather than leaving an R NA.
  site_chr <- as.character(df[[.sed_site_col]])
  site_chr[is.na(site_chr)] <- "nan"
  df$Site <- trimws(site_chr)
  df$Date <- as.Date(df[[.sed_date_col]])
  Period[!is.na(df$Date) & df$Date <= BASELINE_END] <- "Baseline"
  df$Period <- Period
  df$Season <- Season
  df
}
